transport-native-io_uring/src/test/java/io/netty/channel/uring/NativeTest.java

```diff
@@ -270,4 +270,31 @@ public void run() {
             ringBuffer.close();
         }
     }
+
+    @Test
+    public void testUserData() {
+        RingBuffer ringBuffer = Native.createRingBuffer(32);
+        IOUringSubmissionQueue submissionQueue = ringBuffer.ioUringSubmissionQueue();
+        final IOUringCompletionQueue completionQueue = ringBuffer.ioUringCompletionQueue();
+
+        try {
+            // Ensure userdata works with negative and positive values
+            for (int i = Short.MIN_VALUE; i <= Short.MAX_VALUE; i++) {
+                submissionQueue.addWrite(-1, -1, -1, -1, i);
+                assertEquals(1, submissionQueue.submitAndWait());
+                final int expectedData = i;
+                assertEquals(1, completionQueue.process(new IOUringCompletionQueue.IOUringCompletionQueueCallback() {
+                    @Override
+                    public void handle(int fd, int res, int flags, int op, int data) {
+                        assertEquals(-1, fd);
+                        assertTrue(res < 0);
+                        assertEquals(Native.IORING_OP_WRITE, op);
+                        assertEquals(expectedData, data);
+                    }
+                }));
```

Comment:
If encoding and decoding were package-private static methods, ideally placed next to each other in some utility class, the code might be cleaner and we could simplify this test a bit.